runtime/server/x86/frontend/fbank.h

```diff
@@ -0,0 +1,212 @@
+      float left_mel = mel_low_freq + bin * mel_freq_delta,
+      center_mel = mel_low_freq + (bin + 1) * mel_freq_delta,
+      right_mel = mel_low_freq + (bin + 2) * mel_freq_delta;
+      center_freqs_[bin] = InverseMelScale(center_mel);
+      std::vector<float> this_bin(num_fft_bins);
+      int first_index = -1, last_index = -1;
+      for (int i = 0; i < num_fft_bins; i++) {
+        float freq = (fft_bin_width * i);  // Center frequency of this fft
+        // bin.
+        float mel = MelScale(freq);
+        if (mel > left_mel && mel < right_mel) {
+          float weight;
+          if (mel <= center_mel)
+            weight = (mel - left_mel) / (center_mel - left_mel);
+          else
+            weight = (right_mel-mel) / (right_mel-center_mel);
+          this_bin[i] = weight;
+          if (first_index == -1)
+            first_index = i;
+          last_index = i;
+        }
+      }
+      assert(first_index != -1 && last_index >= first_index);
+      bins_[bin].first = first_index;
+      int size = last_index + 1 - first_index;
+      bins_[bin].second.resize(size);
+      for (int i = 0; i < size; i++) {
+        bins_[bin].second[i] = this_bin[first_index + i];
+      }
+    }
+
+    // hamming window
+    hamming_window_.resize(frame_length_);
+    double a = M_2PI / (frame_length-1);
+    for (int i = 0; i < frame_length; i++) {
+        hamming_window_[i] = 0.54 - 0.46*cos(a * i);
+    }
+  }
+
+  void SetUseLog(bool use_log) {
+    use_log_ = use_log;
+  }
+
+  void SetRemoveDcOffset(bool remove_dc_offset) {
+    remove_dc_offset_ = remove_dc_offset;
+  }
+
+  void SetDither(float dither) {
+    dither_ = dither;
+  }
+
+  int NumBins() const { return num_bins_; }
+
+  static inline float InverseMelScale(float mel_freq) {
+    return 700.0f * (expf (mel_freq / 1127.0f) - 1.0f);
+  }
+
+  static inline float MelScale(float freq) {
+    return 1127.0f * logf (1.0f + freq / 700.0f);
+  }
+
+  int UpperPowerOfTwo(int n) const {
+    return static_cast<int>(pow(2, ceil(log(n) / log(2))));
+  }
+
+  // preemphasis
+  void PreEmphasis(float coeff, std::vector<float>* data) const {
+    if (coeff == 0.0) return;
+    for (int i = data->size() - 1; i > 0; i--)
+      (*data)[i] -= coeff * (*data)[i-1];
+    (*data)[0] -= coeff * (*data)[0];
+  }
+
+  // Apply hamming window on data in place
+  void Hamming(std::vector<float>* data) const {
+    assert(data->size() >= hamming_window_.size());
+    for (size_t i = 0; i < hamming_window_.size(); i++) {
+      (*data)[i] *= hamming_window_[i];
+    }
+  }
+
+  // Compute fbank feat, return num frames
+  int Compute(const std::vector<float>& wave,
+              std::vector<std::vector<float> >* feat) {
+    int num_samples = wave.size();
+    if (num_samples < frame_length_) return 0;
+    int num_frames = 1 + ((num_samples - frame_length_) / frame_shift_);
+    feat->resize(num_frames);
+    std::vector<float> fft_real(fft_points_, 0), fft_img(fft_points_, 0);
```

Comment:
Remove `memset`. `fft_real` and `fft.img` are zero initialized!
The definition should be moved to where they are used.

Reply:
it is defined outside the loop, and it is modified for a new frame.

Reply:
Ok, I see. But for `fft_real`, only elements in the interval [frame_length_, fft_points_)
need to be zeroed out.

Reply:
I see, will fix it.